pkg/templatelookup/regular_test.go

```diff
@@ -0,0 +1,108 @@
+	"github.com/kyma-project/lifecycle-manager/pkg/testutils/builder"
+)
+
+func TestFilterTemplate(t *testing.T) {
+	tests := []struct {
+		name     string
+		template templatelookup.ModuleTemplateInfo
+		kyma     *v1beta2.Kyma
+		wantErr  error
+	}{
+		{
+			name: "When TemplateInfo contains Error, Then the output is same as input",
+			template: templatelookup.ModuleTemplateInfo{
+				Err: templatelookup.ErrTemplateNotAllowed,
+			},
+			wantErr: templatelookup.ErrTemplateNotAllowed,
+		},
+		{
+			name: "When ModuleTemplate is internal but Kyma is not, Then result contains error",
+			template: templatelookup.ModuleTemplateInfo{
+				ModuleTemplate: builder.NewModuleTemplateBuilder().
+					WithLabel(shared.InternalLabel, "true").Build(),
+			},
+			kyma: builder.NewKymaBuilder().
+				WithLabel(shared.InternalLabel, "false").
+				Build(),
+			wantErr: templatelookup.ErrTemplateNotAllowed,
+		},
+		{
+			name: "When ModuleTemplate is beta but Kyma is not, Then result contains error",
+			template: templatelookup.ModuleTemplateInfo{
+				ModuleTemplate: builder.NewModuleTemplateBuilder().
+					WithLabel(shared.BetaLabel, "true").Build(),
+			},
+			kyma: builder.NewKymaBuilder().
+				WithLabel(shared.BetaLabel, "false").
+				Build(),
+			wantErr: templatelookup.ErrTemplateNotAllowed,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := templatelookup.FilterTemplate(tt.template, tt.kyma,
+				provider.NewCachedDescriptorProvider()); !errors.Is(got.Err, tt.wantErr) {
+				t.Errorf("FilterTemplate() = %v, want %v", got, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestMarkInvalidChannelSkewUpdate(t *testing.T) {
+	tests := []struct {
+		name                  string
+		moduleTemplate        *templatelookup.ModuleTemplateInfo
+		moduleTemplateVersion string
+		moduleStatus          *v1beta2.ModuleStatus
+		wantErr               error
+	}{
+		{
+			name: "When upgrade version during channel switch, Then result contains no error",
+			moduleTemplate: &templatelookup.ModuleTemplateInfo{
+				ModuleTemplate: &v1beta2.ModuleTemplate{Spec: v1beta2.ModuleTemplateSpec{Channel: "fast"}},
+			},
+			moduleTemplateVersion: "1.1.0",
+			moduleStatus: &v1beta2.ModuleStatus{
+				Channel:  "regular",
+				Version:  "1.0.0",
+				Template: &v1beta2.TrackingObject{TypeMeta: apimetav1.TypeMeta{Kind: "ModuleTemplate"}},
+			},
+			wantErr: nil,
+		}, {
+			name: "When downgrade version during channel switch, Then result contains error",
+			moduleTemplate: &templatelookup.ModuleTemplateInfo{
+				ModuleTemplate: &v1beta2.ModuleTemplate{Spec: v1beta2.ModuleTemplateSpec{Channel: "fast"}},
+			},
+			moduleTemplateVersion: "1.0.0",
+			moduleStatus: &v1beta2.ModuleStatus{
+				Channel:  "regular",
+				Version:  "1.1.0",
+				Template: &v1beta2.TrackingObject{TypeMeta: apimetav1.TypeMeta{Kind: "ModuleTemplate"}},
+			},
+			wantErr: templatelookup.ErrTemplateUpdateNotAllowed,
+		},
+	}
```

Comment:
good point regarding first three conditions, I'll add some cases. regarding the version validation, I think those logic should be removed, if the version validation happens here, it's too later, it should validate when moduletemplate created, which we already did [here](https://github.com/kyma-project/lifecycle-manager/blob/d6c42e009612bd729131af18f9ac1dfab57cb34b/api/v1beta2/moduletemplate_webhook.go#L103), so I would remove those two condition check.

Reply:
Not sure if it should be removed, since the validation is only done in a webhook. We have no guarantee that the webhook is called, do we? In some scenarios webhook might be down, so I would vote for keeping the version validation. but you decide. If you keep the validation, could be worth to move it into a specific pkg to test it there, since it's being used then by the webhook and this templatelookup.

Reply:
I refactored this function to private, and test channel switch update in `GetRegularTemplates` directly, those addtional test mentioned by you most are assertion guard cases, not make too much sense test them specifically.

Reply:
> Not sure if it should be removed, since the validation is only done in a webhook. We have no guarantee that the webhook is called, do we? In some scenarios webhook might be down, so I would vote for keeping the version validation. but you decide. If you keep the validation, could be worth to move it into a specific pkg to test it there, since it's being used then by the webhook and this templatelookup.

no, if moduletemplate webhook is not working, then ModuleTemplate will failed to create or update by API Server, the webhook failurepolicy is configured to Fail
https://github.com/kyma-project/lifecycle-manager/blob/c66ed445c4636086a239f3a4f23b5f6b1e9e060d/config/webhook/manifests.yaml#L14

But I agree we could think refactoring this piece a bit better, currently `IsValidVersionChange` is already shared between webhook and this templatelookup, I think we can also put version validation there. But I suggest address this concern in followup issue.